### ISSUE TITLE
gcov plugin isn't working at all

### ISSUE THREAD

**Issue by wolf99** [ThrowTheSwitch/Ceedling#224]
It seems the gcov plugin isn't working. Well, it is in that it does run. It never finds any of the files though so the resulting report is always empty 😬 

Test project, with incorrect output included, here [https://github.com/wolf99/test-ceedling-gcov](https://github.com/wolf99/test-ceedling-gcov)

**Comment by dpostorivo** [comment]
With Version 0.28.2, I get this on the first run:

```
(venv)veg@meano:~/dev/test-ceedling-gcov$ ceedling gcov:all


Test 'test_five.c'
------------------
Generating runner for test_five.c...
Compiling test_five_runner.c...
Compiling test_five.c...
Compiling unity.c...
Compiling five.c with coverage...
Compiling cmock.c...
Linking test_five.out...
Running test_five.out...

--------------------------
GCOV: OVERALL TEST SUMMARY
--------------------------
TESTED:  1
PASSED:  1
FAILED:  0
IGNORED: 0


---------------------------
GCOV: CODE COVERAGE SUMMARY
---------------------------
five.c Lines executed:100.00% of 2
five.c No branches
five.c No calls
```

and this on the subsequent

```
(venv)veg@meano:~/dev/test-ceedling-gcov$ ceedling gcov:all


Test 'test_five.c'
------------------
Running test_five.out...

--------------------------
GCOV: OVERALL TEST SUMMARY
--------------------------
TESTED:  1
PASSED:  1
FAILED:  0
IGNORED: 0


---------------------------
GCOV: CODE COVERAGE SUMMARY
---------------------------
five.c Lines executed:100.00% of 2
five.c No branches
five.c No calls
```

I can do a PR on your repo to show you what I have after doing `ceedling new` if you want.

What's the output of `ceedling version` and `ruby -v`?

**Comment by wolf99** [comment]
Dammit. `ceedling gcov:all` is for gcov first; I was using `ceedling utils:gcov` 😬. 
The plugin really needs some kind of documentation - I lost an afternoon messing about trying to get this working yesterday. I'd volunteer to do it but obviously I haven't got a clue what I'm at with it.

**Comment by wolf99** [comment]
While I'm here (and to hijack this issue), I notice that in that example project, the coverage report now lists `cmock.c` and `unity.c` files - which according to the defaults.yml `-e` flag should be ignored. On other projects I have recently, I have solved this by adding my own `-e` to `:arguments:` in `project.yml`, but I am unsure why the former one does not prevent this.

**Comment by dpostorivo** [comment]
Yep that broke in by default [this commit](https://github.com/ThrowTheSwitch/Ceedling/commit/44088eb44f2bde8e99dc257e9d1d3af278d08bf1). On my system the single quotes worked, but it did not work for windows. Double quotes alone also work on my system, but I'd rather not go through with the PR until I know it also works on windows. 


If you are on a Windows system can you try removing the single quotes on the lines changed in the commit above in your vendor directory? If it works you can make a PR.

**Comment by wolf99** [comment]
Ah, I see. I'm building on Ubuntu WSL. I tried installing Ruby on windows to attempt this, but like other occasions I've had to investigate MSYS2 (for gcc, gcov, gcovr) I unfortunately don't have the patience or time for that mess 😆. Unless there's some way of using the `ridk` tool to add and use them?

**Comment by wolf99** [comment]
I've been using this in Ubuntu WSL. I tried this on Windows (MSYS2 for gcc & gcov, python pip for gcovr and RubyInstaller for Ruby) and I have not been able to get gcovr to run correctly at all. Tried with just single quotes, just double quotes and both double and single quotes, to no avail.

Always get the output similar:
```bash
C:\Users\me\git\test-ceedling-gcov>ceedling utils:gcov
Creating a detailed html report of gcov results in build/artifacts/gcov/GcovCoverageResults.html...
rake aborted!
Errno::ENOENT: No such file or directory - gcovr -p -b -e "'^vendor.*|^build.*|^test.*|^lib.*'" --html --html-details -r . -o  "build/artifacts/gcov/GcovCoverageResults.html"
C:/Users/me/git/test-ceedling-gcov/vendor/ceedling/lib/ceedling/system_wrapper.rb:45:in ``'
C:/Users/me/git/test-ceedling-gcov/vendor/ceedling/lib/ceedling/system_wrapper.rb:45:in `shell_backticks'
C:/Users/me/git/test-ceedling-gcov/vendor/ceedling/lib/ceedling/tool_executor.rb:68:in `exec'
vendor/ceedling/plugins/gcov/gcov.rake:168:in `block (2 levels) in <top (required)>'
C:/Ruby24-x64/bin/ceedling:23:in `load'
C:/Ruby24-x64/bin/ceedling:23:in `<main>'
Tasks: TOP => utils:gcov
(See full trace by running task with --trace)
ERROR: Ceedling Failed
```

**Comment by dpostorivo** [comment]
It seems gcovr was never installed or never added to path. 


 Btw I have the draft documentation done, but Travis has been behaving oddly for OSX builds and I'm trying to finish debuging that before submitting a PR. You can start to take a look at it [here[(https://github.com/dpostorivo/Ceedling/commit/324f9ac78c8412a031592d6688540910e377dc4f) and give feedback. I'm not sure how long it will take for me to figure out what's up with travis.

**Comment by gruemelmonster** [comment]
For me it is working on windows (with MSYS2):
With the single quotes removed uinty.c and cmock.c get suppressed. I did not expect it to remove the complete path in front of the files though. 

For some reasons it is not working with MINGW64 (but neither with nor without the quotes)

**Comment by wolf99** [comment]
Ok - found time to sort my path out. I had Python's `Scripts` dir on my path but needed to also append `;.PY;.PYC` to the `PATHEXT` system variable and ensure the `gcovr` script filename ended with `.py`.

Once that is done calling `ceedling utils:gcov` still does not work. It results in (note, no single quotes):

```bash
>ceedling utils:gcov
Creating a detailed html report of gcov results in build/artifacts/gcov/GcovCoverageResults.html...
rake aborted!
Errno::ENOENT: No such file or directory - gcovr -p -b -e "^vendor.*|^build.*|^test.*|^lib.*" --html --html-details -r . -o  "build/artifacts/gcov/GcovCoverageResults.html" -e="^Debug.*|^test.*|^build.*|^vendor.*"
C:/Users/Me/git/test-ceedling-gcov/vendor/ceedling/lib/ceedling/system_wrapper.rb:45:in ``'
C:/Users/Me/git/test-ceedling-gcov/vendor/ceedling/lib/ceedling/system_wrapper.rb:45:in `shell_backticks'
C:/Users/Me/git/test-ceedling-gcov/vendor/ceedling/lib/ceedling/tool_executor.rb:68:in `exec'
vendor/ceedling/plugins/gcov/gcov.rake:168:in `block (2 levels) in <top (required)>'
C:/Ruby24-x64/bin/ceedling:23:in `load'
C:/Ruby24-x64/bin/ceedling:23:in `<main>'
Tasks: TOP => utils:gcov
(See full trace by running task with --trace)
ERROR: Ceedling Failed
```

However, running the command _directly_ at the command line results in a different error:

```bash
>gcovr -p -b -e "^vendor.*|^build.*|^test.*|^lib.*" --html --html-details -r . -o  "build/artifacts/gcov/GcovCoverageResults.html" -e="^Debug.*|^test.*|^build.*|^vendor.*"
Traceback (most recent call last):
  File "C:\Users\Me\AppData\Local\Programs\Python\Python36\Scripts\gcovr.py", line 2323, in <module>
    print_html_report(covdata, options.html_details)
  File "C:\Users\Me\AppData\Local\Programs\Python\Python36\Scripts\gcovr.py", line 1698, in print_html_report
    OUTPUT = open(cdata._sourcefile, 'w')
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\Me\\git\\test-ceedling-gcov\\build\\artifacts\\gcov\\GcovCoverageResults.src/five.c.html'
```

Running the same command without the `-o` flag and argument spits the output to the prompt just fine though. I guess this means there's some problem with the output paths?

**Comment by wolf99** [comment]
Seems this is a "known" issue https://github.com/gcovr/gcovr/issues/152.
Furthermore, looking at the code in the installed version, the open merge request that fixes the issue now appears to be out of date 😩

**Comment by davidvanwyk** [comment]
@wolf99 I seem to have managed to get a detailed report going, but it's a bit of a workaround. The post by @AndersHoglund on [gcovr/issues/112](https://github.com/gcovr/gcovr/issues/112) is what prompted the solution. Unfortunately I can't quite figure out how to automate this via the .yml, but hopefully it'll be of some use to you in the interim. 

Here's an example that works on my side, after having identical problems to you and using the utils:gcov output directly as a baseline:

`python C:\Users\davidvw\gcovr\scripts\gcovr.py -p -b -e "'^vendor.*|^build.
*|^test.*|^lib.*'" --html --html-details -r . -k`

Followed by:

`python C:\Users\davidvw\gcovr\scripts\gcovr.py -p -b -e "'^vendor.*|^build.
*|^test.*|^lib.*'" --html --html-details -r . -k -g -o  "build/artifacts/gcov/GcovCoverageResults.html"`

I can't give you any sage wisdom as to why, unfortunately. But hopefully someone more knowledgeable on this than I can come up with a fix from this.

**Comment by Letme** [comment]
If you just list them in 
```
:gcov_report:
:arguments:
  - -g
  - -o build/artifacts/gcov/GcovCoverageResults.html
```
it does not work?

**Comment by davidvanwyk** [comment]
Unfortunately not - still the same error that OP is getting (if trying to do it via ceedling).

**Comment by davidvanwyk** [comment]
An update on this from my side - if I run 

'clobber'
'ceedling gcov:all'
`gcovr.py -p -b -e "^\.launches.*|^\.settings.*|^asm.*|^debug.*|^doc.*|^targetConfigs.*|^test.*|^test-build.*|^tools.*|^vendor.*" --html --html-details -r . -k -g -o build/artifacts/gcov/coverage.html` 

All works as expected. The issue is only present when using utils:gcov to try and run the command.

**Comment by wolf99** [comment]
I'm away from the machine that I set this all up on until the new year so it'tll be a while before I can confirm. Thanks for your input though @davidvanwyk , and happy holidays to all 🎄

**Comment by wolf99** [comment]
I have tried @davidvanwyk attempts on my PC and found that the first (with the two commands) works as required. the latter - just the one command does not work properly. It does output a single html file but the file holds an empty report (i,e, as if there were no files under test).

Why this should be the case I'm unsure.

There is a recent flurry of activity on the gcovr repo after a long period of quietness (I've cast my hat in the ring over there to try to help out) so hopefully any Windows issues can get fixed in the near future.

**Comment by davidvanwyk** [comment]
That is very strange, indeed. 

I noticed the pickup in activity on the gcovr repo and am relieved to see that things are progressing - it's an exceptionally useful tool. If I had the time at the moment I'd certainly put my name forward to assist, but hopefully when I get back into the studying I'll be able to contribute.

**Comment by wolf99** [comment]
The gcov repository is seeing quite a bit of enthusiasm! Some of the fixes for Windows paths have been merged (not in a pip release yet, use the following if you would like to test `> python -m pip install git+https://github.com/gcovr/gcovr.git` appending `--upgrade` if you already have gcovr).

so far this means that gcovr will now build detailed HTML reports. In the process however, the exclude option (`-e`) seems to have broken. There is an issue opened about it here https://github.com/gcovr/gcovr/issues/191

**Comment by nscooling** [comment]
@wolf99 I'm still getting coverage for ```unity.c``` and ```cmock.c``. You mentioned in this thread: 
> I have solved this by adding my own -e to :arguments: in project.yml
Please could you share exactly what it looked like? I haven't been able to get this working through the ```project.yml``` file (using ```gcovr``` from the cli works fine).

**Comment by wolf99** [comment]
@nscooling If you check the gcov plugin's [`defaults.yml`](https://github.com/ThrowTheSwitch/Ceedling/blob/master/plugins/gcov/config/defaults.yml) you will be able to see how the options are set there. Any options you add to your `project.yml` with the same format should override and add to the defaults.

Additionally you can see some examples of configuring gcov options on #94.

Further, if you are running on Windows, `gcovr` has a issues with a few of the options particularly with the exclude option. IIRC _at the time I posted that_, it was a matter of fiddling with the single-quotes vs double-quotes vs both for the exclude option.

If you still are seeing issues, do open a separate issue with a link to a copy of your current `project.yml` if possible 😃

**Comment by nscooling** [comment]
Thanks, I've managed to fix it by modifying the default.yml file. Working on OSX and Linux, so I've done a pull request.

**Comment by psheph** [comment]
I just found this issue today. My goal was to get code coverage reporting set up on our Jenkins server, and I've worked it out locally. I think the changes I had to make could be extended to an additional report setting under the gcov plugin without too much work.

Here are my tool versions:
  Ceedling:: 0.28.3
CException:: 1.3.1.18
     CMock:: 2.4.6.217
     Unity:: 2.4.3.122

I am using the Cobertura plugin to post the xml report file on my Jenkins build page.
To make this work, I had to hack my ceedling setup:

- in ($PROJECT_DIR)/vendor/ceedling/plugins/gcov/lib/gcov_constants.rb, I changed GCOV_ARTIFACTS_FILE to a .xml file extension.
- in ($PROJECT_DIR)/vendor/ceedling/plugins/gcov/config/defaults.yml, I changed the :gcov_post_report_advanced::arguments: to remove --html and --html-details, and add --xml.

In my best case scenario, this would be a new reporting option accessible from the project.yml file.

**Comment by dpostorivo** [comment]
@psheph I do not think this is related to this the original point of this thread and sound like a feature request rather than a bug. The gcov utility has the following help option:

`rake utils:gcov                   # Create gcov code coverage html report (must run ceedling gcov first)`

As you can see from the above, it was always meant to generate an html report, not an xml report.

It would be trivial to add this as an option you can even do the PR if you want.

As a side note @wolf99  is there anything more that needs to be done in ceedling to close this thread? I'd be happy to work on it, but on a quick glance I don't see what the current state of this issue is.

**Comment by psheph** [comment]
@dpostorivo I agree, it's not really related, but I figured out how to handle my problem through a careful reading of this thread. What is the best way to do a feature request? Just open a new issue with [Feature Request] at the front of the title?

Thanks for your collective patience. I'm new to github and unit testing, but since I learned something useful yesterday, I didn't want to wait before giving back to the wider community.

Thanks,
Paul

**Comment by gsmithufl** [comment]
I think the issue with cmock.c and utils.c/h has been fixed in the current version of gcovr through pip. On my WSL Ubuntu system I apt installed gcovr, which is v3.1. After removing that and `pip install gcovr` and refreshing the issue has been resolved.

```
Ceedling:: 0.28.3
CException:: 1.3.1.18
CMock:: 2.4.6.217
Unity:: 2.4.3.122

gcov (Ubuntu 7.4.0-1ubuntu1~18.04) 7.4.0

gcovr 4.1
```

Does this issue still need to be open? Its sort of migrated from its original intent I think.

**Comment by mvandervoord** [comment]
Agreed, @gsmithufl . Thanks everyone!
